### PR TITLE
bugfix for the reconstruction part and the squash function and the margin&reconsturction loss function.

### PR DESCRIPTION
**Pull request by JiaweiWang-AI**
#2 
Get 98% accuracy in 1st epoch now.
Original code is so great and neat. thanks a lot :)
But I found some bugs here.

fixbug1: Use the correct capsule to reconstruct the input image rather than longest capsule. 
fixbug2: Use the squash function to squash the right dimension(unit_size 8) in primary capsule, It have a great impact on the model accuracy. `CapsuleLayer.squash(u, dim=1)`(The digit part squash function seems not right, and the most weird thing is if I squash the right dim which is capsule size 16, model can't be train correctly.)
fixbug3: The margin loss function lack of a square term.
fixbug4: The reconstruction_loss function should minimize the sum of squared differences instead of the mean squared differences.(as the capsule paper said)

### CODE REVIEW

File: capsule_conv_layer.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def __init__(self, in_channels, out_channels):
                                out_channels=out_channels,
                                kernel_size=9, # fixme constant
                                stride=1,
-                               bias=False)
+                               bias=True)
```
Comment (timomernick) [inline capsule_conv_layer.py:22]: good find!  i was experimenting with this myself locally but didn't want to commit either way.  did you measure an accuracy improvement on MNIST with this?

File: capsule_network.py [hunk 2]
```diff
@@ -43,16 +43,18 @@ def __init__(self,
...
+                                   in_channels=num_primary_units,
                                    num_units=num_output_units,
                                    unit_size=output_unit_size,
                                    use_routing=True)
 
         reconstruction_size = image_width * image_height * image_channels
-        self.reconstruct0 = nn.Linear(num_output_units*output_unit_size, (reconstruction_size * 2) / 3)
-        self.reconstruct1 = nn.Linear((reconstruction_size * 2) / 3, (reconstruction_size * 3) / 2)
-        self.reconstruct2 = nn.Linear((reconstruction_size * 3) / 2, reconstruction_size)
+        # self.reconstruct0 = nn.Linear(num_output_units*output_unit_size, (reconstruction_size * 2) / 3)
```
Comment (timomernick) [inline capsule_network.py:53]: are you sure this is correct?  the paper (and other implementations i've seen) seem to say that the reconstruction input is *all* capsules, but with the inactive capsules masked out.

File: capsule_network.py [hunk 3]
```diff
@@ -130,11 +117,11 @@ def reconstruction_loss(self, images, input, size_average=True):
             vutils.save_image(output_image, "reconstruction.png")
         self.reconstructed_image_count += 1
 
-        # The reconstruction loss is the mean squared difference between the input image and reconstructed image.
+        # The reconstruction loss is the sum squared difference between the input image and reconstructed image.
         # Multiplied by a small number so it doesn't dominate the margin (class) loss.
         error = (output - images).view(output.size(0), -1)
         error = error**2
-        error = torch.mean(error, dim=1) * 0.0005
+        error = torch.sum(error, dim=1) * 0.0005
```
Comment (timomernick) [inline capsule_network.py:124]: good find!

File: main.py [hunk 4]
```diff
@@ -46,8 +46,8 @@
 
 conv_inputs = 1
 conv_outputs = 256
-num_primary_units = 8
-primary_unit_size = 32 * 6 * 6  # fixme get from conv2d
+num_primary_units = 32 * 6 * 6
+primary_unit_size = 8  # fixme get from conv2d
```
Comment (timomernick) [inline main.py:50]: i think you're right, i did this part wrong.  let me re-read the paper and consider your change carefully.

File: capsule_network.py [hunk 5]
```diff
@@ -87,29 +89,14 @@ def margin_loss(self, input, target, size_average=True):
...
-        # Stack masked capsules over the batch dimension.
-        masked = torch.stack(all_masked, dim=0)
+    def reconstruction_loss(self, images, input, target, size_average=True):
+        # Use the target to reconstruct input image.
+        # (batch_size, num_output_units, output_unit_size)
+        input = torch.squeeze(input, 3)
+        # (batch_size, num_output_units, 1)
+        target = torch.unsqueeze(target, 2)
+        # (batch_size, output_unit_size, 1)
+        masked = torch.matmul(input.transpose(2,1), target)
```
Comment (timomernick) [inline capsule_network.py:99]: hmm, i don't quite understand this part.  what does 'masked' look like for a typical sample after this matmul?  also see earlier comment, i thought the input was all capsules with all but one masked out -- otherwise the reconstruction would be confused by different classes of digits

**Review by timomernick — CHANGES_REQUESTED** [review]
would you mind splitting things up into small patches?  for example the missing squares in loss term, mean/sum stuff, i can accept immediately.  but the reconstruction change and the transposing of unit size/count needs some more investigation and thought

**Comment by timomernick** [comment]
@JaveyWang i took a couple of your small fixes and pushed them in d2e9aa876793408350e0521648d59990665d0df3

i'm still thinking about your more substantial changes and will leave this PR open.  it would help if you could resolve the conflicts in this PR and reduce it to only the proposed squash and capsule size/count changes.  thanks!

File: capsule_network.py (hunk 5 above)
Comment (JiaweiWang-AI) [reply capsule_network.py:99]: First, there is no doubt to use the target to reconstruction. But my implementation maybe wrong after I rethink about this question. I thought the reconstruction network just need 16 values to reconstruct 10 images, but we know every digit capsule represent 1 entity(digit), so we should use all the 160 values to finish the reconstruction,  and this also prove the fact how to reconstruct 2 overlapped MNIST at a time as the original paper said(Although this can be done by find the two max longest digit capsule and then reconstruct them two times and sum it, not enough neat as aforementioned). So yes, I think you're right. :)
just change the masked part to ' masked = input * target ' and change the number of first layer units to ' num_output_units*output_unit_size ' then it will work.

File: capsule_network.py (hunk 2 above)
Comment (JiaweiWang-AI) [reply capsule_network.py:53]: yes, you're right.  Now I can get 98.9% in first epoch after fix few bug.  But I still wonder why we softmax the wrong dim and squash the wrong dim in routing part can still get such a great result. 
And the most weird thing is that if I make the dim correct this model  would crash sometimes, and get bad accuracy... Hope you can fix it. Thx.

File: capsule_conv_layer.py (hunk 1 above)
Comment (JiaweiWang-AI) [reply capsule_conv_layer.py:22]: Oh, I just think if there is no BN layer in network, the bias should be add up. This strategy is work in traditional convnet. But I don't know if it is useful too in capsnet, I will test it after this model can get better accuracy. It always stuck at about 99.4% acc in my local version.

File: capsule_network.py (hunk 3 above)
Comment (JiaweiWang-AI) [reply capsule_network.py:124]: Thx. Your code is very neat, helpful to me.

**Comment by ghost** [comment]
@JaveyWang  I am also confused with your fixbug2. When I correct the squash dim the trainging result is going to be bad.  Do you have any progress with this?
